src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -167,41 +157,166 @@ internal int EndRead(IAsyncResult asyncResult)
         {
-            if (asyncResult == null)
+            ValidateParameters(buffer, offset, count);
+
+            SslWriteAsync writeAdapter = new SslWriteAsync(_sslState, cancellationToken);
+
+            return WriteAsyncInternal(writeAdapter, buffer, offset, count);
+        }
+
+        private Task WriteAsyncInternal<T>(T writeAdapter, byte[] buffer, int offset, int count)
+            where T : ISslWriteAdapter
+        {
+            _sslState.CheckThrow(authSuccessCheck: true, shutdownCheck: true);
+
+            if (count == 0 && !SslStreamPal.CanEncryptEmptyMessage)
             {
-                throw new ArgumentNullException(nameof(asyncResult));
+                // If it's an empty message and the PAL doesn't support that, we're done.
+                return Task.CompletedTask;
             }
 
-            LazyAsyncResult lazyResult = asyncResult as LazyAsyncResult;
-            if (lazyResult == null)
+            if (Interlocked.Exchange(ref _nestedWrite, 1) == 1)
             {
-                throw new ArgumentException(SR.Format(SR.net_io_async_result, asyncResult.GetType().FullName), nameof(asyncResult));
+                throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, "WriteAsync", "write"));
+            }
+
+            if (count <= _sslState.MaxDataSize)
+            {
+                //Single write
+                Task t = WriteSingleChunk(writeAdapter, buffer, offset, count);
+                if (t.IsCompleted)
+                {
+                    _nestedWrite = 0;
+                    return t;
+                }
+                else
+                {
+                    return ExitWriteAsync(t);
+                }
+            }
+            else
+            {
+                //Loop
+                return WriteAsyncChunked(writeAdapter, buffer, offset, count);
             }
 
-            if (Interlocked.Exchange(ref _nestedWrite, 0) == 0)
+            async Task ExitWriteAsync(Task task)
             {
-                throw new InvalidOperationException(SR.Format(SR.net_io_invalidendcall, "EndWrite"));
+                try
+                {
+                    await task.ConfigureAwait(false);
+                }
+                catch (Exception e)
+                {
+                    _sslState.FinishWrite();
+
+                    if (e is IOException)
+                    {
+                        throw;
+                    }
+
+                    throw new IOException(SR.net_io_write, e);
+                }
+                finally
+                {
+                    _nestedWrite = 0;
+                }
             }
+        }
 
-            // No "artificial" timeouts implemented so far, InnerStream controls timeout.
-            lazyResult.InternalWaitForCompletion();
+        private async Task WaitForWriteIOSlot<T>(T writeAdapter, Task lockTask, byte[] buffer, int offset, int count)
+            where T : ISslWriteAdapter
```

Comment:
Just to be safe, this should probably be "where T : ISslWriteAdapter, struct"

Reply:
changed, (needs to be struct, ISslWriteAdapter but same diff :P )